File: Lecture1-C/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>

char* cardtype;
char str[20];
char strnum[2];
long cardnumber;
int length = 0;
int leftsum = 0;
int rightsum = 0;
int newsum = 0;
int mark = 0;
int newmult = 0;
long get_cardnumber(void);
int get_cardlength(long cardnumber);

int main(void) 
{
    // get user input for credit card number
    long cardnumber = get_cardnumber();

    // check length of card
    int length = get_cardlength(cardnumber);
    
    // convert the int to a string
    sprintf(str, "%ld", cardnumber);

    // Luhn's Algorithms
    // multiply by 2 for each 2 digits from the right
    for (int i = length-1; i >= 0; i-=2)
    {
        int num = str[i] - 48;
        int mult = num * 2;
        while (mult != 0)
        {
            rightsum += mult % 10;
            mult /= 10;
        }
    }

    // get the sum of digits that weren't multiply by 2
    for (int k = length; k >= 0; k-=2)
    {
        int num = str[k] - 48;
        leftsum += num;
    }

    // if mod result is 0 then the card is valid
    int sums = leftsum + rightsum;
    if (sums % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // check length
    if (length <= 10)
    {
        printf("INVALID\n");
        return 0;
    }

    if (str[0] == '3' && (str[1] == '4' || str[1] == '7'))
    {
        cardtype = "AMEX";
    }
    else if (str[0] == '5' && (str[1] == '1' || str[1] == '2' || str[1] == '3' || str[1] == '4' || str[1] == '5'))
    {
        cardtype = "MASTERCARD";
    }
    else if (str[0] == '4')
    {
        cardtype = "VISA";
    }
    else
    {
        cardtype = "INVALID";
    }

    // print the output
    printf("%s\n", cardtype);

}

long get_cardnumber(void)
{
    do 
    {
        cardnumber = get_long("Number: ");
    }
    while (cardnumber < 0);
    return cardnumber;
}

int get_cardlength(long cardnumber)
{
    do 
    {
        cardnumber = cardnumber/10;
        length += 1;
    }
    while (cardnumber >= 10);
    return length;
}

File: Lecture6-Python/pset6/sentimental-credit/credit.py
# TODO
import cs50
import math

def main():
    number = cs50.get_string("Number: ")
    
    if number != None and len(number) <= 10:
        return "INVALID"
    
    if (number != None and is_card(number) == True):
        if int(number[0]) == 3 and (int(number[1]) == 4 or int(number[1]) == 7):
            return "AMEX"
        elif int(number[0]) == 4:
            return "VISA"
        elif int(number[0]) == 5 and (int(number[1]) == 1 or int(number[1]) == 2 or int(number[1]) == 3 or int(number[1]) == 4 or int(number[1]) == 5):
            return "MASTERCARD"
        else:
            return "INVALID"
    else:
        return "INVALID"

def is_card(number):
    rightsum = get_right_sum(number)
    leftsum = get_left_sum(number)
    if ((rightsum + leftsum) % 10) == 0:
        return True
    return False

def get_right_sum(number):
    rightsum = 0
    idx = len(number)-2
    while idx >= 0:
        num = 2* int(number[idx])
        while num != 0:
            rightsum += num % 10
            num = math.floor(num / 10)
        idx -= 2
    return rightsum

def get_left_sum(number):
    leftsum = 0
    idx = len(number)-1
    while idx >= 0:
        leftsum += int(number[idx])
        idx -= 2
    return leftsum
    
print(main())